### PR TITLE
PP-5498: add query parameters to get transaction endpoint

### PR DESCRIPTION
**Pull request by jankowiakmaria**
* add `transaction_type` and `parent_external_id` optional query parameters
* extended transaction Dao method
* updated (and added new) integration test

### CODE REVIEW

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java [hunk 1]
```diff
@@ -20,7 +21,9 @@
 
     private static final String FIND_TRANSACTION_BY_EXTERNAL_ID_AND_GATEWAY_ACCOUNT_ID = "SELECT * FROM transaction " +
             "WHERE external_id = :externalId " +
-            "and gateway_account_id = :gatewayAccountId";
+            "AND gateway_account_id = :gatewayAccountId " +
+            "AND (:transactionType::transaction_type is NULL OR type = :transactionType::transaction_type) " +
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: Makes me worried why this change didn't break existing pacts. For `a transaction with created state exist` we don't currently pass transactionType parameter and as the query should now be `transaction_type is NULL` (which is not currently) I am expecting pact test to break unless I am missing something ?

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java [hunk 2]
```diff
@@ -20,7 +21,9 @@
...
+            "AND (:parentExternalId is NULL OR parent_external_id = :parentExternalId)";
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: Are we making it mandatory to provide `parentExternalId` for get one refund transaction ?

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java [hunk 3]
```diff
@@ -120,11 +123,13 @@ public TransactionDao(Jdbi jdbi) {
         this.jdbi = jdbi;
     }
 
-    public Optional<TransactionEntity> findTransactionByExternalIdAndGatewayAccountId(String externalId, String gatewayAccountId) {
+    public Optional<TransactionEntity> findTransactionByExternalIdAndGatewayAccountId(String externalId, String gatewayAccountId, TransactionType transactionType, String parentTransactionExternalId) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: Method name could change as it's purpose for transactions with more filters

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: Looks like we don't have a pact test for `get-one-transaction`

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 2 above)
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: I don't think it needs to be mandatory on ledger side. If it's not valid, it won't return any results.

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 3 above)
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: yeah, wasn't sure how to name it.

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 1 above)
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: `:transactionType::transaction_type is NULL` checks whether the passed parameter is null, not the value of the column in db.
For Payment we don't currently pass value for transaction type when we retrieve it in public api from ledger. (It will be done separately)

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 2 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: This could break if `:parentExternalId` is an empty string which could be ignored ?

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: I prefer constructing query if value is available. Not a great fan of using `NULL is NULL` condition but fine with it too.

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 3 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: May be 
```suggestion
    public Optional<TransactionEntity> findTransaction(String externalId, String gatewayAccountId, TransactionType transactionType, String parentTransactionExternalId) {
```

File: src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java (hunk 2 above)
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: Should it be an empty string?

Comment (kbottla) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: This is the usual pattern for non-mandatory fields until it is different in this case 

https://github.com/alphagov/pay-ledger/blob/bc62be837a511359e961dcad08293e4c6947a943/src/main/java/uk/gov/pay/ledger/transaction/search/common/TransactionSearchParams.java#L271

Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java]: I don't think it's required in this case. Public api doesn't specify this parameter at all when calling ledger, so it shouldn't be an issue.
If we discover otherwise we can update the code to handle the scenario.
